libnmstate/error.py

```diff
@@ -0,0 +1,96 @@
+#
+# Copyright 2019 Red Hat, Inc.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 2 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+try:
+    PermissionError
+except NameError:
+    # Python 2 does not have PermissionError.
+    class PermissionError(Exception):
+        pass
+
+
+class NmstateError(Exception):
+    """
+    The base exception of libnmstate.
+    """
+    def __init__(self, msg):
+        super(NmstateError, self).__init__(msg)
+        self._msg = msg
+
+    @property
+    def msg(self):
+        """
+        Error message of specified NmstateError exception.
+        """
+        return self._msg
+
+    def __str__(self):
+        return self._msg
+
+
+class NmstateExternalRequirementError(NmstateError):
+    """
+    Nmstate requires external tools installed and/or started for desired state.
+    """
+    pass
+
+
+class NmstateInvalidArgumentError(NmstateError, ValueError):
+    """
+    Exception happens at pre-apply check, user should resubmit the amended
+    desired state. Example:
+        * JSON/YAML syntax issue.
+        * Nmstate schema issue.
+        * Invalid value of desired property, like bond missing slave.
+    """
+    pass
+
+
+class NmstatePermissionDenyError(NmstateError, PermissionError):
```

Comment:
IMHO it could just be `NmstatePermissionError` because error already implies that the permission is wrong. If you insist on the deny part, it should be `Denied` IMHO.

Reply:
Changed to `NmstateNoSupportError`.

Reply:
IMHO it would be better to use `NmstateNotSupportedError` to have the same tense for the names.